sklearn/base.py

```diff
@@ -60,6 +64,14 @@ def clone(estimator, *, safe=True):
     return different results from the original estimator. More details can be
     found in :ref:`randomness`.
     """
+    if hasattr(estimator, "__sklearn_clone__") and not inspect.isclass(estimator):
+        return estimator.__sklearn_clone__()
+    return _clone_parametrized(estimator, safe=safe)
+
+
+def _clone_parametrized(estimator, *, safe=True):
```

Comment:
Will this benefit from being public? Or will those who need it be able to use `super`?

Reply:
It's not safe for a third party estimator to call `clone` from its `__sklearn_clone__`. Yes it could call `BaseEstimator.__sklearn_clone__` but then it needs to import `BaseEstimator` (or copy-paste).

Having a public `clone_default_implementation` or `clone_parametrized` wouldn't hurt...? But I don't know how important it is to support non-use of BaseEstimator.

Reply:
I thought we're not caring about not having to inherit from BaseEstimator as much any more?

Reply:
maybe let's keep it like this and see is someone complains that they can't use it?

Reply:
I agree with keeping this PR as is and develop a solution if a third party estimator developer requires it.